Implement APIs for palveluväylä

solita-antti-mottonen opened the pull request.

solita-antti-mottonen committed:
Implement APIs for palveluväylä

Implements: AE-1959

solita-antti-mottonen committed:
Remove checking stderr when running LibreOffice

LibreOffice prints out warning to stderr on newer MacOS versions. The
warning doesn't seem to be anything serious, and there doesn't seem to
be any solutions for it available. Exit code is still checked for
errors.

Juholei commented on etp-backend/src/main/clj/solita/etp/api/palveluvayla.clj:
```diff
@@ -1,18 +1,43 @@
+                                           {{:keys [accept-language]} :header} :parameters
+                                           {{:keys [id]} :path}                :parameters}]
+                                       (let [language-preference-order (if accept-language
+                                                                         (->> accept-language
```
Tämä voisi olla oma funktionsa

Juholei commented on etp-backend/src/main/clj/solita/etp/api/palveluvayla.clj:
```diff
@@ -1,18 +1,43 @@
+                                                                              (map first))
+                                                                         ["fi" "sv"])]
+                                         (api-response/pdf-response ; Return the first language version that exists if any
+                                           (some identity (->> language-preference-order
```
Olisiko tämä kieliversion päättely parempi olla service-kerroksessa?

Juholei commented on etp-backend/src/main/clj/solita/etp/service/energiatodistus_pdf.clj:
```diff
@@ -697,14 +697,16 @@
   (when-let [{:keys [allekirjoitusaika] :as complete-energiatodistus}
-             (complete-energiatodistus-service/find-complete-energiatodistus db whoami id)]
+             (-> (complete-energiatodistus-service/find-complete-energiatodistus db whoami id)
+                 (#(when (or (= (-> % :perustiedot :kieli) 2)
```
Tätä voisi selkiyttää jotenkin

Juholei commented on etp-backend/src/test/clj/solita/etp/api/palveluvayla_test.clj:
```diff
@@ -0,0 +1,138 @@
+(def palveluvayla-basepath "/api/palveluvayla/energiatodistukset")
+
+(t/deftest test-palveluvayla-api
+  (let [laatija-id (first (keys (test-data.laatija/generate-and-insert! 1)))
```
Tähän testisetupiin voisi kommenteilla kertoa, että mitä tässä oikein tehdään, koska tätä on niin paljon ja hälyisän näkyistä koodia, että siitä ei pikavilkaisulla oikein ymmärrä mitä tapahtuu.

Juholei commented on etp-backend/src/test/clj/solita/etp/api/palveluvayla_test.clj:
```diff
@@ -0,0 +1,138 @@
+          (t/is (= 200 (:status response)))
+          (t/is (= 3 (count body)))
+          (t/is (= #{todistus-2013-fi-id todistus-2013-sv-id todistus-2013-multilingual-id} (set (map :id body))))
+          (schema-tools.coerce/coerce body [schema.energiatodistus/Energiatodistus2013] schema-tools.coerce/json-coercion-matcher)))
```
Tuottaako tämä coerce-kutsu lisäarvoa tässä siihen nähden, että rajapinta nähdäkseni jo itsessään validoi datan tätä samaa skeemaa vasten?

solita-antti-mottonen replied:
Oikeastihan tässä pitäisi varmistaa datan sisältö, mutta sitä en jaksanut alkaa vääntämään. Tämä nyt kuitenkin varmistaa että rajapinnassa on asetettu oikea skeema, ja koska oikeat id tarkistetaan yläpuolella niin data on todennäköisesti se mitä haluttiin.

solita-antti-mottonen committed:
Requested changes

Create a service for Palveluväylä to make the codebase a bit clearer.
Separate parsing languages into a function. Clean up the test code a
bit. Comment language discrimination in pdf search.

Juholei commented on etp-backend/src/main/clj/solita/etp/service/palveluvayla.clj:
```diff
@@ -0,0 +1,41 @@
+   (service.energiatodistus/find-energiatodistus-any-laatija db id))
+  ([id db version]
+   (-> (service.energiatodistus/find-energiatodistus db id)
+       (#(drop-wrong-version % version)))))
```
Tarvitseeko tässä enää tätä anonyymiin funktioon käärimistä?

Juholei commented on etp-backend/src/main/clj/solita/etp/service/palveluvayla.clj:
```diff
@@ -0,0 +1,41 @@
+  "Return the first language version that exists if any. If language-preference-order is not given, [fi sv] is used."
+  [id language-preference-order db aws-s3-client]
+  (some identity (->> (or language-preference-order ["fi" "sv"])
+                      (map #(service.energiatodistus-pdf/find-energiatodistus-pdf db
```
Tässä voisi ehkä tuon mapin korvata jollakin ei-laiskalla funktiolla (ihan vain mapv?), jotta energiatodistuksen haku blob storagesta tapahtuu varmasti ennakoitavaan aikaan koodissa.

(some identity) voisi myös olla threading-makrossa?

solita-antti-mottonen replied:
Ideana oli hakea niistä vain eka joka löytyy. Clojurehan ei niin tee, mutta ajattelin että kommunikoi paremmin haetun toiminnallisuuden jos niin tekee. 

(some identity) toimii myö makrossa, mutta taas minusta näin päin tuntuu kommunikoivan tarkoituksen paremmin.

solita-antti-mottonen committed:
Add a feature flag for Palveluväylä API

solita-antti-mottonen committed:
Simplify threading macro

Juholei commented on etp-backend/src/test/clj/solita/etp/api/palveluvayla_test.clj:
```diff
@@ -0,0 +1,150 @@
+        (t/is (= 200 (:status response)))
+        (t/is (= todistus-2013-fi-id (:id body)))
+        (schema-tools.coerce/coerce body schema.energiatodistus/EnergiatodistusForAnyLaatija schema-tools.coerce/json-coercion-matcher)))
+    (t/testing "Fetching energiatodistus by id returns it when fetching correct version"
```
Laita näiden testing-lohkojen väliin tyhjät rivit, niin erottaa paremmin missä yksi testicase päättyy ja toinen alkaa.

Juholei commented on etp-backend/src/test/clj/solita/etp/api/palveluvayla_test.clj:
```diff
@@ -0,0 +1,150 @@
+            body (j/read-value (:body response) j/keyword-keys-object-mapper)]
+        (t/is (= 200 (:status response)))
+        (t/is (= 6 (count body)))
+        (t/is (= #{todistus-2013-fi-id todistus-2018-fi-id todistus-2013-sv-id todistus-2018-sv-id todistus-2013-multilingual-id todistus-2018-multilingual-id} (set (map :id body))))
```
Laittaisin näin monen alkion setissä jokaisen omalle rivilleen kuten myös tuon mihin sitä testissä vertaillaan. Erottuu sitten paremmin myös se mitä tähän vertaillaan.

solita-antti-mottonen committed:
Improve formatting
